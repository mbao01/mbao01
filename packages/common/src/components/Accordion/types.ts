import * as AccordionPrimitive from "@radix-ui/react-accordion";

export type AccordionProps = React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Root>;

export type AccordionContentProps = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Content
>;

export type AccordionItemProps = React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Item>;

export type AccordionTriggerProps = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Trigger
>;
