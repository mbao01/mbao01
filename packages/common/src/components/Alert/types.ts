import { type VariantProps } from "../../libs";
import { getAlertClasses } from "./constants";

export type AlertProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getAlertClasses>;

export type AlertTitleProps = React.HTMLAttributes<HTMLHeadingElement>;

export type AlertDescriptionProps = React.HTMLAttributes<HTMLDivElement>;
