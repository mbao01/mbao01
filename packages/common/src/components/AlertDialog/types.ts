import * as AlertDialogPrimitive from "@radix-ui/react-alert-dialog";
import { type VariantProps } from "../../libs";
import { getButtonClasses } from "../Button/constants";
import {
  getDialogContentClasses,
  getDialogDescriptionClasses,
  getDialogFooterClasses,
  getDialogHeaderClasses,
  getDialogOverlayClasses,
  getDialogTitleClasses,
} from "../Dialog/constants";

export type AlertDialogProps = React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Root>;

export type AlertDialogOverlayProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Overlay
> &
  VariantProps<typeof getDialogOverlayClasses>;

export type AlertDialogContentProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Content
> &
  VariantProps<typeof getDialogContentClasses>;

export type AlertDialogHeaderProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogHeaderClasses>;

export type AlertDialogFooterProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogFooterClasses>;

export type AlertDialogTitleProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Title
> &
  VariantProps<typeof getDialogTitleClasses>;

export type AlertDialogDescriptionProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Description
> &
  VariantProps<typeof getDialogDescriptionClasses>;

export type AlertDialogActionProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Cancel
> &
  VariantProps<typeof getButtonClasses>;

export type AlertDialogCancelProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Cancel
> &
  VariantProps<typeof getButtonClasses>;
