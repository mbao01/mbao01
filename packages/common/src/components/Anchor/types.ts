import { VariantProps } from "../../libs";
import { getAnchorClasses } from "./constant";

export type AnchorProps = React.AnchorHTMLAttributes<HTMLAnchorElement> & {
  isExternal?: boolean;
} & VariantProps<typeof getAnchorClasses>;
