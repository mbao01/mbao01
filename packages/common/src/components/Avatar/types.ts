import * as AvatarPrimitive from "@radix-ui/react-avatar";
import { type VariantProps } from "../../libs";
import {
  getAvatarClasses,
  getAvatarFallbackClasses,
  getAvatarImageClasses,
} from "./constants";

export type AvatarProps = Omit<
  React.ComponentPropsWithoutRef<typeof AvatarPrimitive.Root>,
  "asChild" | "size"
> &
  VariantProps<typeof getAvatarClasses>;

export type AvatarImageProps = Omit<
  React.ComponentPropsWithoutRef<typeof AvatarPrimitive.Image>,
  "asChild"
> &
  VariantProps<typeof getAvatarImageClasses>;

export type AvatarFallbackProps = Omit<
  React.ComponentPropsWithoutRef<typeof AvatarPrimitive.Fallback>,
  "asChild"
> &
  VariantProps<typeof getAvatarFallbackClasses>;
