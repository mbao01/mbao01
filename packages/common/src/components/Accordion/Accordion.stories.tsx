import type { Meta, StoryContext, StoryFn, StoryObj } from "@storybook/react";
import { ChevronDownIcon } from "@radix-ui/react-icons";
import { Accordion } from "./Accordion";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const withAccordion = (_: StoryFn, context: StoryContext<any>) => {
  const items = [
    {
      value: "item-1",
      trigger: "Is it accessible?",
      content: "Yes. It adheres to the WAI-ARIA design pattern.",
    },
    {
      value: "item-2",
      trigger: "Is it styled?",
      content: "Yes. It comes with default styles that matches the other components aesthetic.",
    },
    {
      trigger: "Is it animated?",
      content: "Yes. It's animated by default, but you can disable it if you prefer.",
      value: "item-3",
    },
  ];

  const disabled = (context.args as { disabled: boolean }).disabled;

  return (
    <Accordion {...context.args} className="w-[320px]">
      {items.map((item) => (
        <Accordion.Item value={item.value} key={item.value}>
          <Accordion.Trigger disabled={item.value === "item-3" ? disabled : false}>
            {item.trigger}
            <ChevronDownIcon className="h-4 w-4 shrink-0 transition-transform duration-200" />
          </Accordion.Trigger>
          <Accordion.Content>{item.content}</Accordion.Content>
        </Accordion.Item>
      ))}
    </Accordion>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Accordion,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
    // backgrounds: [{ name: 'dark background', value: '#000', default: true }],
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withAccordion],
} satisfies Meta<typeof Accordion>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    type: "single",
    collapsible: true,
  },
};

export const ExpandMultipleItems: Story = {
  args: {
    type: "multiple",
  },
};

export const SingleExpandedItem: Story = {
  args: {
    type: "single",
    defaultValue: "item-2",
    collapsible: true,
  },
};

export const MultipleExpandedItems: Story = {
  args: {
    type: "multiple",
    defaultValue: ["item-1", "item-3"],
  },
};

export const NonCollapsible: Story = {
  args: {
    type: "single",
    collapsible: false,
  },
};

export const DisabledLastItem: Story = {
  args: {
    type: "single",
    collapsible: true,
    disabled: true,
  },
};
