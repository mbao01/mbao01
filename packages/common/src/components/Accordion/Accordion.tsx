"use client";

import * as React from "react";
import * as AccordionPrimitive from "@radix-ui/react-accordion";
import type {
  AccordionProps,
  AccordionContentProps,
  AccordionItemProps,
  AccordionTriggerProps,
} from "./types";
import { cn } from "../../utilities";
import {
  getAccordionContentClasses,
  getAccordionItemClasses,
  getAccordionTriggerClasses,
} from "./constants";

const Accordion = (props: AccordionProps) => (
  <AccordionPrimitive.Root {...props} />
);

const AccordionItem = React.forwardRef<
  React.ElementRef<typeof AccordionPrimitive.Item>,
  AccordionItemProps
>(({ className, ...props }, ref) => (
  <AccordionPrimitive.Item
    ref={ref}
    className={cn(getAccordionItemClasses(), className)}
    {...props}
  />
));
AccordionItem.displayName = "AccordionItem";

const AccordionTrigger = React.forwardRef<
  React.ElementRef<typeof AccordionPrimitive.Trigger>,
  AccordionTriggerProps
>(({ className, children, ...props }, ref) => (
  <AccordionPrimitive.Header className="flex">
    <AccordionPrimitive.Trigger
      ref={ref}
      className={cn(getAccordionTriggerClasses(), className)}
      {...props}
    >
      {children}
    </AccordionPrimitive.Trigger>
  </AccordionPrimitive.Header>
));
AccordionTrigger.displayName = AccordionPrimitive.Trigger.displayName;

const AccordionContent = React.forwardRef<
  React.ElementRef<typeof AccordionPrimitive.Content>,
  AccordionContentProps
>(({ className, children, ...props }, ref) => (
  <AccordionPrimitive.Content
    ref={ref}
    className={cn(getAccordionContentClasses())}
    {...props}
  >
    <div className={cn("pb-4 pt-0", className)}>{children}</div>
  </AccordionPrimitive.Content>
));
AccordionContent.displayName = AccordionPrimitive.Content.displayName;

Accordion.Item = AccordionItem;
Accordion.Trigger = AccordionTrigger;
Accordion.Content = AccordionContent;

export { Accordion };
