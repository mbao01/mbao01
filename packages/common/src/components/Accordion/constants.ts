import { cva } from "../../libs";

export const getAccordionItemClasses = cva("border-b");

export const getAccordionTriggerClasses = cva(
  "flex flex-1 items-center justify-between py-4 text-sm font-medium transition-all [&:hover:not(:disabled)]:underline disabled:opacity-50 [&[data-state=open]>svg]:rotate-180"
);

export const getAccordionContentClasses = cva(
  "overflow-hidden text-sm data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
);
