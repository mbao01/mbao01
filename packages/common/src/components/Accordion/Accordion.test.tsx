import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it } from "vitest";
import { Accordion } from "./Accordion";
import { type AccordionProps } from "./types";

describe("Accordion", () => {
  const items = [
    {
      value: "item-1",
      trigger: "Is it accessible?",
      content: "Yes. It adheres to the WAI-ARIA design pattern.",
      disabled: false,
    },
    {
      value: "item-2",
      trigger: "Is it styled?",
      content: "Yes. It comes with default styles that matches the other components aesthetic.",
      disabled: false,
    },
    {
      trigger: "Is it animated?",
      content: "Yes. It's animated by default, but you can disable it if you prefer.",
      value: "item-3",
      disabled: true,
    },
  ];
  const renderAccordion = (props: AccordionProps = { type: "single" }) => {
    return render(
      <Accordion {...props}>
        {items.map((item) => (
          <Accordion.Item value={item.value} key={item.value}>
            <Accordion.Trigger disabled={item.disabled}>{item.trigger}</Accordion.Trigger>
            <Accordion.Content>{item.content}</Accordion.Content>
          </Accordion.Item>
        ))}
      </Accordion>
    );
  };

  it("renders an accordion", () => {
    const { asFragment } = renderAccordion();

    items.forEach((item) => {
      expect(screen.getByRole("heading", { name: item.trigger })).toBeInTheDocument();
      expect(screen.getByRole("button", { name: item.trigger })).toHaveAttribute(
        "aria-expanded",
        "false"
      );
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("should expand and collapse an accordion item", async () => {
    const user = userEvent.setup();

    renderAccordion({
      type: "single",
      collapsible: true,
      defaultValue: "item-3",
    });

    expect(screen.getByRole("button", { name: items[0].trigger })).toHaveAttribute(
      "aria-expanded",
      "false"
    );
    expect(screen.getByRole("button", { name: items[1].trigger })).toHaveAttribute(
      "aria-expanded",
      "false"
    );
    expect(screen.getByRole("button", { name: items[2].trigger })).toHaveAttribute(
      "aria-expanded",
      "true"
    );

    await user.click(screen.getByRole("button", { name: items[1].trigger }));

    expect(screen.getByRole("button", { name: items[0].trigger })).toHaveAttribute(
      "aria-expanded",
      "false"
    );
    expect(screen.getByRole("button", { name: items[1].trigger })).toHaveAttribute(
      "aria-expanded",
      "true"
    );
    expect(screen.getByRole("button", { name: items[2].trigger })).toHaveAttribute(
      "aria-expanded",
      "false"
    );

    await user.click(screen.getByRole("button", { name: items[1].trigger }));

    expect(screen.getByRole("button", { name: items[0].trigger })).toHaveAttribute(
      "aria-expanded",
      "false"
    );
    expect(screen.getByRole("button", { name: items[1].trigger })).toHaveAttribute(
      "aria-expanded",
      "false"
    );
    expect(screen.getByRole("button", { name: items[2].trigger })).toHaveAttribute(
      "aria-expanded",
      "false"
    );
  });
});
