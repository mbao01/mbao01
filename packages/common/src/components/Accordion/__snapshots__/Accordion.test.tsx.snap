// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion > renders an accordion 1`] = `
<DocumentFragment>
  <div
    data-orientation="vertical"
  >
    <div
      class="border-b"
      data-orientation="vertical"
      data-state="closed"
    >
      <h3
        class="flex"
        data-orientation="vertical"
        data-state="closed"
      >
        <button
          aria-controls="radix-«r1»"
          aria-expanded="false"
          class="flex flex-1 items-center justify-between py-4 text-sm font-medium transition-all [&:hover:not(:disabled)]:underline disabled:opacity-50 [&[data-state=open]>svg]:rotate-180"
          data-orientation="vertical"
          data-radix-collection-item=""
          data-state="closed"
          id="radix-«r0»"
          type="button"
        >
          Is it accessible?
        </button>
      </h3>
      <div
        aria-labelledby="radix-«r0»"
        class="overflow-hidden text-sm data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
        data-orientation="vertical"
        data-state="closed"
        hidden=""
        id="radix-«r1»"
        role="region"
        style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
      />
    </div>
    <div
      class="border-b"
      data-orientation="vertical"
      data-state="closed"
    >
      <h3
        class="flex"
        data-orientation="vertical"
        data-state="closed"
      >
        <button
          aria-controls="radix-«r3»"
          aria-expanded="false"
          class="flex flex-1 items-center justify-between py-4 text-sm font-medium transition-all [&:hover:not(:disabled)]:underline disabled:opacity-50 [&[data-state=open]>svg]:rotate-180"
          data-orientation="vertical"
          data-radix-collection-item=""
          data-state="closed"
          id="radix-«r2»"
          type="button"
        >
          Is it styled?
        </button>
      </h3>
      <div
        aria-labelledby="radix-«r2»"
        class="overflow-hidden text-sm data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
        data-orientation="vertical"
        data-state="closed"
        hidden=""
        id="radix-«r3»"
        role="region"
        style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
      />
    </div>
    <div
      class="border-b"
      data-orientation="vertical"
      data-state="closed"
    >
      <h3
        class="flex"
        data-orientation="vertical"
        data-state="closed"
      >
        <button
          aria-controls="radix-«r5»"
          aria-expanded="false"
          class="flex flex-1 items-center justify-between py-4 text-sm font-medium transition-all [&:hover:not(:disabled)]:underline disabled:opacity-50 [&[data-state=open]>svg]:rotate-180"
          data-orientation="vertical"
          data-radix-collection-item=""
          data-state="closed"
          disabled=""
          id="radix-«r4»"
          type="button"
        >
          Is it animated?
        </button>
      </h3>
      <div
        aria-labelledby="radix-«r4»"
        class="overflow-hidden text-sm data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
        data-orientation="vertical"
        data-state="closed"
        hidden=""
        id="radix-«r5»"
        role="region"
        style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
      />
    </div>
  </div>
</DocumentFragment>
`;
